File: zip_utils.py
import os
import pandas as pd
import geopandas as gpd
from shapely.geometry import Point

# Default file paths
DEFAULT_ZIP_GEO_FILE = os.path.join(os.path.dirname(__file__), 'constant_data', 'wa_washington_zip_codes_geo.min.json')
DEFAULT_ZIP_POPULATION_FILE = os.path.join(os.path.dirname(__file__), 'constant_data', 'washington_zip_populations.csv')

_zip_geo_df = None

def load_zip_codes(zip_file_path=None):
    """
    Load zip code boundaries from GeoJSON file.
    Returns a GeoDataFrame with zip code polygons.
    
    Args:
        zip_file_path (str, optional): Path to the zip code GeoJSON file.
                                      If None, uses default path.
    """
    global _zip_geo_df
    
    if _zip_geo_df is None:
        if zip_file_path is None:
            zip_file_path = DEFAULT_ZIP_GEO_FILE
        
        if not os.path.exists(zip_file_path):
            print(f"Warning: Zip code file not found at {zip_file_path}")
            return None
            
        try:
            print(f"Loading zip code boundaries from {zip_file_path}...")
            _zip_geo_df = gpd.read_file(zip_file_path)
            print(f"Loaded {len(_zip_geo_df)} zip code boundaries")
            return _zip_geo_df
        except Exception as e:
            print(f"Error loading zip code file: {e}")
            return None


def get_zip_code(lon, lat):
    """
    Get zip code for a given longitude and latitude using spatial join.
    
    Args:
        lon (float): Longitude coordinate
        lat (float): Latitude coordinate
        
    Returns:
        str: Zip code if found, None otherwise
    """
    # Handle missing coordinates
    if pd.isna(lon) or pd.isna(lat):
        return None
    
    # Load zip code boundaries
    if _zip_geo_df is None:
        raise Exception("Zip code boundaries not loaded. Load first with load_zip_codes().")
    
    try:
        # Create a Point geometry for the coordinates
        point = Point(lon, lat)
        
        # Create a GeoDataFrame with the point
        point_gdf = gpd.GeoDataFrame(
            [{'geometry': point}], 
            crs='EPSG:4326'  # WGS84 coordinate system
        )
        
        # Perform spatial join to find which zip code contains the point
        result = gpd.sjoin(
            point_gdf, 
            _zip_geo_df, 
            how='left', 
            predicate='within'
        )
        
        # Extract zip code if found
        if len(result) > 0 and 'ZCTA5CE10' in result.columns:
            zip_code = result.iloc[0]['ZCTA5CE10']
            return str(zip_code) if pd.notna(zip_code) else None
        
        return None
        
    except Exception as e:
        print(f"Error getting zip code for coordinates ({lon}, {lat}): {e}")
        return None

File: src/zip_utils.ts
import * as fs from 'fs';
import * as path from 'path';
import { fileURLToPath } from 'url';
import booleanPointInPolygon from '@turf/boolean-point-in-polygon';
import { point } from '@turf/helpers';

// Get __dirname equivalent in ES modules
const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

// Default file paths
const DEFAULT_ZIP_GEO_FILE = path.join(__dirname, '..', 'constant_data', 'wa_washington_zip_codes_geo.min.json');
const DEFAULT_ZIP_POPULATION_FILE = path.join(__dirname, '..', 'constant_data', 'washington_zip_populations.csv');

interface ZipCodeFeature {
  type: 'Feature';
  properties: {
    ZCTA5CE10: string;
    [key: string]: any;
  };
  geometry: {
    type: 'Polygon' | 'MultiPolygon';
    coordinates: any;
  };
}

interface ZipGeoData {
  type: 'FeatureCollection';
  features: ZipCodeFeature[];
}

let _zipGeoData: ZipGeoData | null = null;

/**
 * Load zip code boundaries from GeoJSON file.
 * Returns zip code data with polygons.
 */
export function loadZipCodes(zipFilePath?: string): ZipGeoData | null {
  if (_zipGeoData !== null) {
    return _zipGeoData;
  }

  const filePath = zipFilePath || DEFAULT_ZIP_GEO_FILE;

  if (!fs.existsSync(filePath)) {
    console.log(`Warning: Zip code file not found at ${filePath}`);
    return null;
  }

  try {
    console.log(`Loading zip code boundaries from ${filePath}...`);
    const fileContent = fs.readFileSync(filePath, 'utf-8');
    _zipGeoData = JSON.parse(fileContent) as ZipGeoData;
    console.log(`Loaded ${_zipGeoData.features.length} zip code boundaries`);
    return _zipGeoData;
  } catch (e) {
    console.log(`Error loading zip code file: ${e}`);
    return null;
  }
}

/**
 * Get zip code for a given longitude and latitude using point-in-polygon.
 */
export function getZipCode(lon: number | null | undefined, lat: number | null | undefined): string | null {
  // Handle missing coordinates
  if (lon === null || lon === undefined || lat === null || lat === undefined) {
    return null;
  }

  if (isNaN(lon) || isNaN(lat)) {
    return null;
  }

  // Load zip code boundaries
  if (_zipGeoData === null) {
    throw new Error('Zip code boundaries not loaded. Load first with loadZipCodes().');
  }

  try {
    // Create a Point for the coordinates
    const pt = point([lon, lat]);

    // Find which zip code contains the point
    for (const feature of _zipGeoData.features) {
      if (booleanPointInPolygon(pt, feature as any)) {
        const zipCode = feature.properties.ZCTA5CE10;
        return zipCode ? String(zipCode) : null;
      }
    }

    return null;
  } catch (e) {
    console.log(`Error getting zip code for coordinates (${lon}, ${lat}): ${e}`);
    return null;
  }
}
